pkg/scheduler/queue/user_queues.go

```diff
@@ -28,298 +27,350 @@ type querier struct {
-		q.users = q.users[:ix]
+	for ix := len(q.tenantQuerierState.tenantIDOrder) - 1; ix >= 0 && q.tenantQuerierState.tenantIDOrder[ix] == ""; ix-- {
+		q.tenantQuerierState.tenantIDOrder = q.tenantQuerierState.tenantIDOrder[:ix]
 	}
 }
 
 // Returns existing or new queue for user.
 // MaxQueriers is used to compute which queriers should handle requests for this user.
 // If maxQueriers is <= 0, all queriers can handle this user's requests.
 // If maxQueriers has changed since the last call, queriers for this are recomputed.
-func (q *queues) getOrAddQueue(userID string, maxQueriers int) *list.List {
-	// Empty user is not allowed, as that would break our users list ("" is used for free spot).
-	if userID == "" {
+func (q *queues) getOrAddTenantQueue(tenantID string, maxQueriers int) *list.List {
+	tenant := q.tenantQuerierState.getOrAddTenant(tenantID, maxQueriers)
+	if tenant == nil {
 		return nil
 	}
+	queue := q.tenantQueues[tenantID]
 
-	if maxQueriers < 0 {
-		maxQueriers = 0
-	}
-
-	uq := q.userQueues[userID]
-
-	if uq == nil {
-		uq = &userQueue{
+	if queue == nil {
+		queue = &userQueue{
 			requests: list.New(),
-			seed:     util.ShuffleShardSeed(userID, ""),
-			index:    -1,
 		}
-		q.userQueues[userID] = uq
-
-		// Add user to the list of users... find first free spot, and put it there.
-		for ix, u := range q.users {
-			if u == "" {
-				uq.index = ix
-				q.users[ix] = userID
-				break
-			}
-		}
-
-		// ... or add to the end.
-		if uq.index < 0 {
-			uq.index = len(q.users)
-			q.users = append(q.users, userID)
-		}
-	}
-
-	if uq.maxQueriers != maxQueriers {
-		uq.maxQueriers = maxQueriers
-		uq.queriers = shuffleQueriersForUser(uq.seed, maxQueriers, q.sortedQueriers, nil)
+		q.tenantQueues[tenantID] = queue
 	}
 
-	return uq.requests
+	return queue.requests
 }
 
 // Finds next queue for the querier. To support fair scheduling between users, client is expected
 // to pass last user index returned by this function as argument. If there was no previous
 // last user index, use -1.
 func (q *queues) getNextQueueForQuerier(lastUserIndex int, querierID string) (*list.List, string, int, error) {
-	uid := lastUserIndex
-
-	// Ensure the querier is not shutting down. If the querier is shutting down, we shouldn't forward
-	// any more queries to it.
-	if info := q.queriers[querierID]; info == nil || info.shuttingDown {
-		return nil, "", uid, ErrQuerierShuttingDown
+	nextTenantID, nextTenantIndex, err := q.tenantQuerierState.getNextTenantIDForQuerier(lastUserIndex, querierID)
+	if err != nil || nextTenantID == "" {
+		return nil, nextTenantID, nextTenantIndex, err
 	}
 
-	for iters := 0; iters < len(q.users); iters++ {
-		uid = uid + 1
+	tenantQueue := q.tenantQueues[nextTenantID]
+	return tenantQueue.requests, nextTenantID, nextTenantIndex, nil
+}
 
-		// Don't use "mod len(q.users)", as that could skip users at the beginning of the list
-		// for example when q.users has shrunk since last call.
```

Comment:
Would be good to keep this comment - this is an important detail.

Reply:
I figured it was old because I couldn't figure out how someone would use mod here instead of standard iteration over the entire list?

Reply:
If the size of `q.users` never changed, then doing something like `uid = (uid + 1) % len(q.users)` would be a way to ensure we wrapped around to the beginning of the list when `uid` moved past the end of the list.

Reply:
done